faiss/gpu/impl/RaftIVFPQ.cu

```diff
@@ -0,0 +1,564 @@
+                    &(raft_knn_index.value()),
+                    raft::make_device_matrix_view<float, uint32_t>(
+                            centroids.data(), numLists_, dim_));
+        } else {
+            /// No reconstruct needed since the centers are already in float32
+            // The FlatIndex keeps its data in float32, so we can merely
+            // reference it
+            auto centroids = gpuData->getVectorsFloat32Ref();
+
+            raft::neighbors::ivf_pq::helpers::set_centers(
+                    raft_handle,
+                    &(raft_knn_index.value()),
+                    raft::make_device_matrix_view<float, uint32_t>(
+                            centroids.data(), numLists_, dim_));
+        }
+    } else {
+        DeviceTensor<float, 2, true> centroids(
+                resources_,
+                makeSpaceAlloc(AllocType::FlatData, space_, stream),
+                {getNumLists(), getDim()});
+
+        // Otherwise, we need to reconstruct all vectors from the index and copy
+        // them to the GPU, in order to have access as needed for residual
+        // computation
+        auto vecs = std::vector<float>(getNumLists() * getDim());
+        quantizer->reconstruct_n(0, quantizer->ntotal, vecs.data());
+
+        centroids.copyFrom(vecs, stream);
+
+        raft::neighbors::ivf_pq::helpers::set_centers(
+                raft_handle,
+                &(raft_knn_index.value()),
+                raft::make_device_matrix_view<float, uint32_t>(
+                        centroids.data(), numLists_, dim_));
+    }
+
+    setPQCentroids_();
+}
+
+/// Return the list indices of a particular list back to the CPU
+std::vector<idx_t> RaftIVFPQ::getListIndices(idx_t listId) const {
+    FAISS_ASSERT(raft_knn_index.has_value());
+    const raft::device_resources& raft_handle =
+            resources_->getRaftHandleCurrentDevice();
+    auto stream = raft_handle.get_stream();
+
+    idx_t listSize = getListLength(listId);
+
+    std::vector<idx_t> vec(listSize);
+
+    // fetch the list indices ptr on host
+    idx_t* list_indices_ptr;
+
+    raft::update_host(
+            &list_indices_ptr,
+            const_cast<idx_t**>(
+                    raft_knn_index.value().inds_ptrs().data_handle()) +
+                    listId,
+            1,
+            stream);
+    raft_handle.sync_stream();
+
+    raft::update_host(vec.data(), list_indices_ptr, listSize, stream);
+    raft_handle.sync_stream();
+
+    return vec;
+}
+
+/// Performs search when we are already given the IVF cells to look at
+/// (GpuIndexIVF::search_preassigned implementation)
+void RaftIVFPQ::searchPreassigned(
+        Index* coarseQuantizer,
+        Tensor<float, 2, true>& vecs,
+        Tensor<float, 2, true>& ivfDistances,
+        Tensor<idx_t, 2, true>& ivfAssignments,
+        int k,
+        Tensor<float, 2, true>& outDistances,
+        Tensor<idx_t, 2, true>& outIndices,
+        bool storePairs) {
+    // TODO: Fill this in!
```

Comment:
We could achieve this by calling https://github.com/rapidsai/raft/blob/0c75bb1985c1fe9177cf18e9c42d071d7115af7b/cpp/include/raft/neighbors/detail/ivf_pq_search.cuh#L840C7-L840C22

We would probably need to add a public interface to `raft::neighbros::ivf_pq::helpers`

Reply:
Leaving this open for discussion. If `searchPreassigned` is not of top priority, we can postpone this for now.

Reply:
I think we should definitely include this in a follow-up.